File: Lecture 1 - C/Problem Set 1/Cash/cash.c
/*
    PROBLEM: CASH (STANDARD C VERSION)

    Write a program that:
    - Prompts user for change owed in cents.
    - Calculates minimum number of coins needed.
    - No CS50 library used.
*/

#include <stdio.h>

// Function prototype
int get_change_owed(void);

int main(void)
{
    int change = get_change_owed();
    int coins = 0;

    coins += change / 25;
    change %= 25;

    coins += change / 10;
    change %= 10;

    coins += change / 5;
    change %= 5;

    coins += change;

    printf("%d\n", coins);

    return 0;
}

// Get input using scanf
int get_change_owed(void)
{
    int change;

    do
    {
        printf("Change owed: ");
        scanf("%d", &change);
    }
    while (change < 0);

    return change;
}

/*
---------------- ALGORITHM ----------------

1. Ask user for change owed in cents.
2. Ensure value is non-negative.
3. Use greedy strategy:
      - Take as many 25¢ coins as possible.
      - Then 10¢ coins.
      - Then 5¢ coins.
      - Remaining are 1¢ coins.
4. Count total coins used.
5. Print total.
6. End.

Time Complexity: O(1)
Space Complexity: O(1)
*/

File: Lecture 6 - Python/Problem Set 6/Cash/cash.py
"""
PROBLEM: CASH (GREEDY ALGORITHM)

Write a program that:
1. Prompts the user for the amount of change owed (in dollars).
2. Converts that amount into cents.
3. Calculates the minimum number of coins required.
4. Uses U.S. coin denominations:
      - 25 cents (quarters)
      - 10 cents (dimes)
      - 5 cents (nickels)
      - 1 cent (pennies)
5. Prints the total number of coins used.
"""


# Keep asking user until valid positive number is entered
while True:
    try:
        change = float(input("Change Owed: "))

        # Ensure change is greater than 0
        if change > 0:
            break
        else:
            print("Give me number above 0")

    # Handle invalid (non-numeric) input
    except ValueError:
        print("Give me number above 0")


# Convert dollars to cents to avoid floating point errors
cents = round(change * 100)


# Calculate number of quarters (25 cents)
quarters = cents // 25
cents %= 25


# Calculate number of dimes (10 cents)
dimes = cents // 10
cents %= 10


# Calculate number of nickels (5 cents)
nickels = cents // 5
cents %= 5


# Remaining cents are pennies
pennies = cents


# Total number of coins
total = quarters + dimes + nickels + pennies

print(total)


"""
---------------- ALGORITHM ----------------

1. Repeat until valid input:
      a. Prompt user for change owed.
      b. Convert input to float.
      c. If input > 0 → continue.
      d. Otherwise → print error message.

2. Convert dollars to cents:
      cents = round(change * 100)

3. Use greedy approach:
      a. quarters = cents // 25
      b. cents = cents % 25
      c. dimes = cents // 10
      d. cents = cents % 10
      e. nickels = cents // 5
      f. cents = cents % 5
      g. pennies = remaining cents

4. Add total coins.
5. Print result.

Time Complexity: O(1)
Space Complexity: O(1)
"""